# ForBio Phylogenomics

A collection of tutorials to be covered in the [ForBio Phylogenomics](https://www.forbio.uio.no/events/courses/2021/phylogenomics.html) course.

## Prepration

* [Requirements](requirements.md)<br>A list of all installations required for the tutorials

* [Getting Started](getting_started.md)<br>General instructions on the use of these tutorials

## Week 1

#### Monday

* [Introduction into the course](week1_day1_morning/README.md)<br>An introduction to the course and the first step of the phylogenetic analyses we will build upon (no exercises)

* [Sources of incongruence & Contamination screening](week1_day1_afternoon/README.md)<br>General lectures on inconrguence and a tutorial on the detection of contamination in datasets using a metabarcoding-based approach.

#### Tuesday

* [Paralogy screening after orthology determination](week1_day2_morning/README.md)<br>A tutorial on the detection of paralogy in datasets using a single trees, bootstrap support and reference databases.

* [Missing data and low phylogenetic signal](week1_day2_afternoon/README.md)<br>A tutorial on reducing missingness and detecting low phylogenetic signal in the data.

#### Wednesday

* [Determination of rogue taxa](week1_day3_morning/README.md)<br>A tutorial on the detection of rogue taxa in datasets using different approaches.

* [Model misspecification and partinioning](week1_day3_afternoon/README.md)<br>A tutorial on reducing model misspecifications by using partitioning schemes.

#### Thursday

* [Saturation](week1_day4_morning/README.md)<br>A tutorial on the detection of saturation in phylogenomic datasets.

* [Base composition heterogeneity](week1_day4_afternoon/README.md)<br>A tutorial on the detection of base composition heterogeneity in phylogenomic datasets.

#### Friday

* [Branch length heterogeneity 1](week1_day5_morning/README.md)<br>A tutorial on the visualization of branch length heterogeneity in phylogenomic datasets.

* [Branch length heterogeneity 2](week1_day5_afternoon/README.md)<br>A tutorial on the detection of branch length heterogeneity in phylogenomic datasets.



## Week 2

#### Monday

* [Bayesian Phylogenetic Inference](bayesian_phylogeny_inference/README.md)<br>A tutorial on Bayesian inference of time-calibrated phylogenies

* [Phylogenetic Divergence-Time Estimation](divergence_time_estimation/README.md)<br>A tutorial on phylogenetic divergence-time estimation with with fossils

#### Tuesday

* [Maximum-Likelihood Species-Tree Inference](ml_species_tree_inference/README.md)<br>A tutorial on maximum-likelihood species-tree inference based on gene trees

* [Bayesian Species-Tree Inference](bayesian_species_tree_inference/README.md)<br>A tutorial on Bayesian inference of time-calibrated species trees

* [Species-Tree Inference with SNP Data](species_tree_inference_with_snp_data/README.md)<br>A tutorial on the inference of species trees from SNP data based on quartets

* [Divergence-Time Estimation with SNP Data](divergence_time_estimation_with_snp_data/README.md)<br>A tutorial on Bayesian divergence-time estimation with SNP data

#### Wednesday

* [Whole-Genome Alignment](XXX)<br>A tutorial on repeat masking and whole-genome alignment of assembled genomes

* [Bayesian Analysis of Species Networks](bayesian_analysis_of_species_networks/README.md)<br>A tutorial on Bayesian phylogenetic analysis of species networks

#### Thursday

* [Analysis of Introgression with SNP Data](analysis_of_introgression_with_snp_data/README.md)<br>A tutorial on the analysis of hybridization and introgression with SNP data

* [Analysis of Introgression with Chromosome-Length Alignments](analysis_of_introgression_with_chromosome_length_alignments/README.md)<br>A tutorial on the analysis of hybridization and introgression with whole-chromosome alignments
<!-- XXX TODO: This might need a substantial update, ideally with addition of AIM -->

#### Friday

* [Inference from Simulated Data](inference_from_simulated_data/README.md)<br>A tutorial on coalescent simulations of genomic data and inferences from simulated data
<!-- This will need to be written from scratch -->
