# Requirements

Most analyses of the tutorials of the ForBio Phylogenomics course will be conducted using the [Saga server](https://documentation.sigma2.no/hpc_machines/saga.html) provided by [Sigma2](https://www.sigma2.no), where the required tools are already be installed and available through modules. There are a few exceptions, however, where the use of Graphical User Interface (GUI) programs will be required or beneficial; these should therefore be installed locally.

## Graphical User Interface programs

<!--* **AliView:** To visualize sequence alignments, the software [AliView](http://www.ormbunkar.se/aliview/) ([Larsson 2014](https://academic.oup.com/bioinformatics/article/30/22/3276/2391211)) is recommended. The installation of AliView is described at [http://www.ormbunkar.se/aliview/](http://www.ormbunkar.se/aliview/) and should be possible on all operating systems.
-->
* **A console program:** A console is essential for work on servers like Saga. Unless you have an old Windows computer, a console should already be installed on your computer, such as the Terminal program on MacOS or Linux or the Windows Console on Windows. All of these programs should be sufficient for the course; however, on the latter system, a more versatile alternative to Windows Console might be an emulator such as [Cmder](https://cmder.app).

* **SeaView:** The program [SeaView](http://evomics.org/resources/software/bioinformatics-software/seaview/) is a graphical multiple sequence alignment editor developed by Manolo Gouy. SeaView is able to read and write various alignment formats (NEXUS, MSF, CLUSTAL, FASTA, PHYLIP, MASE). It allows one to manually edit the alignment, and also to run DOT-PLOT or MSA programs to locally improve the alignment. Download the program [here](https://doua.prabi.fr/software/seaview).

* **AliView:** In addition to SeaView, we are going to use a second alignment viewer, namely [AliView](http://www.ormbunkar.se/aliview/), developed by Anders Larssen. Like SeaView, AliView allows manual editing and translation from nucleotide to amino acid format. Versions compiled for MacOS, Linux, and Windows are available on [http://ormbunkar.se/aliview](http://ormbunkar.se/aliview/#DOWNLOAD).

* **BEAST2:** The BEAST2 package, including BEAUti, BEAST2 itself, TreeAnnotator, and other tools can be downloaded from the BEAST2 website [https://www.beast2.org](https://www.beast2.org). As all these programs are written in Java, compilation is not required, and all programs should work on Mac OS X, Linux, and Windows. I recommend downloading, where possible, the program versions that include the Java Runtime Environment, which may prevent conflicts with Java versions that may already be installed on your machine.<br>

* **Tracer:** The program [Tracer](http://beast.community/tracer) greatly facilitates the inspection of output from Bayesian analyses such as those done with BEAST2. It is a GUI program that therefore can not be used on Saga, but it is easy to install on your local computer. Input files for Tracer will thus need to be downloaded from Saga. Executables of Tracer for MacOS, Linux, and Window can be found on [https://github.com/beast-dev/tracer/releases](https://github.com/beast-dev/tracer/releases). Download the file [Tracer.v1.7.2.dmg](https://github.com/beast-dev/tracer/releases/download/v1.7.2/Tracer.v1.7.2.dmg) if your local computer is running MacOS, [Tracer_v1.7.2.tgz](https://github.com/beast-dev/tracer/releases/download/v1.7.2/Tracer_v1.7.2.tgz) if it is running Linux, and [Tracer_v1.7.2.tgz](https://github.com/beast-dev/tracer/releases/download/v1.7.2/Tracer_v1.7.2.tgz) if it is running Windows.

* **FigTree:** The program [FigTree](http://tree.bio.ed.ac.uk/software/figtree/) is a very intuitive and useful tool for the visualization and (to a limited extent) manipulation of phylogenies encoded in [Newick](http://evolution.genetics.washington.edu/phylip/newicktree.html) format. Being a GUI program, FigTree can not be run on Saga, but needs to be installed and used on your local computer. Input files for FigTree will thus need to be downloaded from Saga. Executables of FigTree for Mac OS X, Linux, and Windows are provided on [https://github.com/rambaut/figtree/releases](https://github.com/rambaut/figtree/releases). Download the file [FigTree.v1.4.4.dmg](https://github.com/rambaut/figtree/releases/download/v1.4.4/FigTree.v1.4.4.dmg) if your local computer is running MacOS, [FigTree_v1.4.4.tgz](https://github.com/rambaut/figtree/releases/download/v1.4.4/FigTree_v1.4.4.tgz) if it is running Linux, and [FigTree.v1.4.4.zip](https://github.com/rambaut/figtree/releases/download/v1.4.4/FigTree.v1.4.4.zip) if it is running Windows.

* **PAUP\*:** The software [PAUP\*](http://paup.phylosolutions.com) is a general-utility program for phylogenetic inference into which the SVDQuartets method, allowing the inference of species trees from SNP data, has been implemented. A command-line version of PAUP\* is installed on Saga, but there is also a Graphical User Interface (GUI) version, which may be easier to use for those not familiar with the program. If you prefer to use the GUI version of PAUP\*, you would need to install it on your local computer, and you would need download the input file for PAUP\* from Saga. Unfortunately, this GUI version does not run on MacOS 10.15 (Catalina) or newer, but on other systems, the program can be installed using the instructions and precompiled versions available on [http://phylosolutions.com/paup-test/](http://phylosolutions.com/paup-test/).

* **RStudio:** [RStudio](https://www.rstudio.com/products/rstudio/) is an integrated development environment (IDE) for [R](https://www.r-project.org/). It includes a console, syntax-highlighting editor that supports direct code execution, as well as tools for plotting, history, debugging and workspace management. RStudio is available in open source and commercial editions and runs on the desktop (Windows, Mac, and Linux) or in a browser connected to RStudio Server or RStudio Server Pro (Debian/Ubuntu, Red Hat/CentOS, and SUSE Linux). Download the program [RStudio Desktop](https://www.rstudio.com/products/rstudio/download/). You have to choose the free version, which is not for the server. You will also have to install [R 3.0.1+](https://cran.rstudio.com/) on your computer before you install R Studio. Choose the OS appropriate for your computer.
